src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,502 @@
+            // validation rules are in place, that should not be possible.
+            if (result.Streams?.Count == 1)
+            {
+                // do not return a result, but set up a subscription
+                var stream = result.Streams!.Single().Value;
+                // note that this may immediately trigger some notifications
+                var disposer = stream.Subscribe(new Observer(this, messageId, _options.DisconnectAfterErrorEvent, _options.DisconnectAfterAnyError));
+                try
+                {
+                    if (Subscriptions.CompareExchange(messageId, dummyDisposer, disposer))
+                    {
+                        disposer = null;
+                    }
+                }
+                finally
+                {
+                    disposer?.Dispose();
+                }
+            }
+            else if (result.Executed)
+            {
+                await SendSingleResultAsync(message, result);
+            }
+            else
+            {
+                await SendErrorResultAsync(message, result);
+            }
+        }
+        catch (OperationCanceledException) when (CancellationToken.IsCancellationRequested)
+        {
+            throw;
+        }
+        catch (ExecutionError error)
+        {
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            await SendErrorResultAsync(message, error);
+        }
+        catch (Exception ex)
+        {
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            var error = await HandleErrorDuringSubscribeAsync(message, ex);
+            await SendErrorResultAsync(message, error);
+        }
+    }
+
+    /// <summary>
+    /// Creates an <see cref="ExecutionError"/> for an unknown <see cref="Exception"/>.
+    /// </summary>
+    protected virtual Task<ExecutionError> HandleErrorDuringSubscribeAsync(OperationMessage message, Exception ex)
+        => Task.FromResult<ExecutionError>(new UnhandledError("Unable to set up subscription for the requested field.", ex));
+
+    /// <summary>
+    /// Sends a single result to the client for a query or mutation request, along with a notice
+    /// that it was the last result in the event stream.
+    /// </summary>
+    protected virtual async Task SendSingleResultAsync(OperationMessage message, ExecutionResult result)
+    {
+        await SendDataAsync(message.Id!, result);
+        await SendCompletedAsync(message.Id!);
+    }
+
+    /// <summary>
+    /// Sends an execution error to the client during set-up of a GraphQL request (typically subscription).
+    /// The default implementation calls <see cref="SendErrorResultAsync(string, ExecutionError)"/>.
+    /// </summary>
+    protected virtual Task SendErrorResultAsync(OperationMessage message, ExecutionError executionError)
+        => SendErrorResultAsync(message.Id!, executionError);
```

Comment:
```suggestion
    /// <summary>
    /// Sends an execution error to the client during set-up of a GraphQL request (typically subscription).
    /// The default implementation calls <see cref="SendErrorResultAsync(string, ExecutionResult)"/>.
    /// </summary>
    protected virtual Task SendErrorResultAsync(OperationMessage message, ExecutionError executionError)
        => SendErrorResultAsync(message.Id!, new ExecutionResult { Errors = new ExecutionErrors { executionError } });
```

To simplify things.

Reply:
I disagree; it seems simpler to me the way it is now.

Reply:
I suggested that all overloads call last abstract method without intermediate hops - plain delegating instead of nested.

Reply:
Ok